File: Pulse/BasicPulsetvOSPlayer/objc/BasicPulsetvOSPlayer/ListViewController.m
//
//  ListViewController.m
//  BasicPulsetvOSPlayer
//
//  Created by Carlos Ceja on 2/18/20.
//  Copyright © 2020 Carlos Ceja. All rights reserved.
//

#import <AppTrackingTransparency/AppTrackingTransparency.h>

#import "ListViewController.h"

#import "BCOVPulseVideoItem.h"
#import "ViewController.h"


@interface ListViewController () <UITableViewDelegate, UITableViewDataSource>

@property (nonatomic, weak) IBOutlet UITableView *tableView;

@property (nonatomic, strong) NSArray <BCOVPulseVideoItem *> *videoItems;

@end


@implementation ListViewController

- (void)viewDidLoad
{
    [super viewDidLoad];

    if (@available(tvOS 14, *))
    {
        __weak typeof(self) weakSelf = self;
        [ATTrackingManager requestTrackingAuthorizationWithCompletionHandler:^(ATTrackingManagerAuthorizationStatus status) {
            __strong typeof(weakSelf) strongSelf = weakSelf;
            dispatch_async(dispatch_get_main_queue(), ^{
                // Tracking authorization completed. Start loading ads here.
                [strongSelf videoLibrary];
            });
        }];
    }
    else
    {
        [self videoLibrary];
    }
}

#pragma mark Misc

- (NSArray<NSDictionary *> *)jsonVideoObjects
{
    NSError *jsonError;
    NSString *path  = [[NSBundle mainBundle] pathForResource:@"Library" ofType:@"json"];
    NSArray<NSDictionary *> *jsonObjects = [NSJSONSerialization JSONObjectWithData:[[NSData alloc] initWithContentsOfFile:path] options:0 error:&jsonError];
    
    assert(jsonError == nil);
    return jsonObjects;
}

- (void)videoLibrary
{
    if (!self.videoItems)
    {
        // Parse and add each video in the JSON array to our video library
        NSMutableArray *videos = [NSMutableArray array];
        
        for (NSDictionary *jsonObject in [self jsonVideoObjects])
        {
            [videos addObject:[BCOVPulseVideoItem initWithDictionary:jsonObject]];
        }
        
        self.videoItems = [NSArray arrayWithArray:videos];
    }

    [self.tableView reloadData];
}

#pragma mark UITableViewDataSource

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return self.videoItems.count;
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section
{
    return @"Basic Pulse tvOS Player";
}

- (void)tableView:(UITableView *)tableView willDisplayHeaderView:(UIView *)view forSection:(NSInteger)section
{
    ((UITableViewHeaderFooterView *) view).contentView.backgroundColor = UIColor.darkGrayColor;
    ((UITableViewHeaderFooterView *) view).textLabel.textColor = UIColor.whiteColor;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"BasicCell" forIndexPath:indexPath];
    
    BCOVPulseVideoItem *item = self.videoItems[indexPath.item];
    
    cell.textLabel.text = item.title ?: @"";
    cell.textLabel.textColor = UIColor.blackColor;
    
    NSString *subtitle = item.category ?: @"";
    
    cell.detailTextLabel.text = [NSString stringWithFormat:@"%@ %@", subtitle, [item.tags componentsJoinedByString:@", "] ?: @""];
    cell.detailTextLabel.textColor = UIColor.grayColor;
    
    return cell;
}

#pragma mark Navigation

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    if ([segue.identifier isEqualToString:@"ViewControllerSegue"])
    {
        NSIndexPath *indexPath = [self.tableView indexPathForSelectedRow];
        BCOVPulseVideoItem *videoItem = self.videoItems[indexPath.item];

        ViewController *destinationVC = segue.destinationViewController;
        destinationVC.videoItem = videoItem;
    }
}

@end

File: Pulse/BasicPulsetvOSPlayer/swift/BasicPulsetvOSPlayer/ListViewController.swift
//
//  ListViewController.swift
//  BasicPulsetvOSPlayer
//
//  Created by Carlos Ceja on 3/13/20.
//  Copyright © 2020 Brightcove. All rights reserved.
//

import AppTrackingTransparency
import UIKit

class ListViewController: UIViewController
{
    
    @IBOutlet private weak var tableView: UITableView!
    lazy fileprivate var videoItems: [BCOVPulseVideoItem] = {
        
        var _videoItems = [BCOVPulseVideoItem]()
        
        if let path = Bundle.main.path(forResource: "Library", ofType: "json")
        {
            do
            {
                let data = try Data(contentsOf: URL(fileURLWithPath: path), options: .mappedIfSafe)
                let jsonResult = try JSONSerialization.jsonObject(with: data, options: .mutableLeaves)
                if let jsonResult = jsonResult as? [Dictionary<String, AnyObject>]
                {
                    for item in jsonResult
                    {
                        _videoItems.append(BCOVPulseVideoItem.staticInit(dictionary: item))
                    }
                }
            }
            catch
            {
                 print("LisViewController Debug - Error retrieving library")
            }
        }
        
        return _videoItems
    }()

    override func viewDidLoad()
    {
        super.viewDidLoad()

        if #available(tvOS 14, *) {
            ATTrackingManager.requestTrackingAuthorization { (status: ATTrackingManager.AuthorizationStatus) in
                print("Tracking authorization completed.")
            }
        }
    }

    
    // MARK: Navigation
    override func prepare(for segue: UIStoryboardSegue, sender: Any?)
    {
        if (segue.identifier == "ViewControllerSegue")
        {
            let indexPath = tableView.indexPathForSelectedRow
            let videoItem = videoItems[indexPath?.item ?? 0]
            
            let destinationVC = segue.destination as? ViewController
            destinationVC?.videoItem = videoItem
        }
    }
}


// MARK: - UITableViewDelegate, UITableViewDataSource

extension ListViewController: UITableViewDelegate, UITableViewDataSource
{
    func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String?
    {
        return "Basic Pulse tvOS Player"
    }
    
    func tableView(_ tableView: UITableView, willDisplayHeaderView view: UIView, forSection section: Int)
    {
        (view as! UITableViewHeaderFooterView).contentView.backgroundColor = UIColor.darkGray
        (view as! UITableViewHeaderFooterView).textLabel?.textColor = UIColor.white
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int
    {
        return videoItems.count
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell
    {
        let item = videoItems[indexPath.item]

        let cell = tableView.dequeueReusableCell(withIdentifier: "BasicCell", for: indexPath)

        cell.textLabel?.text = item.title ?? ""
        cell.textLabel?.textColor = UIColor.black
        cell.detailTextLabel?.text = "\(item.category ?? "") \(item.tags?.joined(separator: ", ") ?? "")"
        cell.detailTextLabel?.textColor = UIColor.gray
        
        return cell
    }

}
